### PR TITLE
Dev/186 New Markdown Editor: Bugs in ExamQuestions.

### PR DESCRIPTION
**Pull request by nyngwang**
Under construction. Don't review for now.

### CODE REVIEW

File: src/app/exam/options/ExamOptions.js [hunk 1]
```diff
@@ -9,117 +9,107 @@ import {examService} from "../../../services/services";
...
-                            </div>
-                            <TitleLine title="Danger Zone"/>
-                            <DangerZone onDangerButtonClick={() => setShowDeleteExamModal(true)}
-                                        header="Delete this exam"
-                                        description="Once you delete an exam, there is no going back. Please be certain."
-                                        buttonName="Delete Exam"/>
+    return <>
+        <div className="exam-options">
+            <div className="font-poppins" style={{paddingTop: "20px", paddingBottom: "150px"}}>
+                {/* TODO: Refactor into scss. */}
```
Comment (Johnny850807) [inline src/app/exam/options/ExamOptions.js:61]: do it right now

File: src/app/exam/questions/ExamQuestions.js [hunk 2]
```diff
@@ -101,93 +84,18 @@ const ExamQuestions = () => {
...
+    if (!exam || !questions)
+        return <Spinner/>;
 
     return (
         <div className="exam-questions">
-            <ExamInPageNavigationBar
-                currentURL={currentURL}
-                examName={currentExam.name}
-                examId={examId}/>
+            {/* TODO: refactor into scss. */}
```
Comment (Johnny850807) [inline src/app/exam/questions/ExamQuestions.js:98]: do it right now

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

File: src/app/exam/options/ExamOptions.js [hunk 3]
```diff
@@ -9,117 +9,99 @@ import {examService} from "../../../services/services";
...
-    });
+        if (!exam)
+            examService.getExam(examId)
+                       .then(exam => {
+                           setExamName(exam.name);
+                           setStartTime(formatDate(exam?.startTime));
+                           setEndTime(formatDate(exam?.endTime));
+                           return exam;
+                       })
+                       .then(setExam);
```
Comment (Wally5077) [inline src/app/exam/options/ExamOptions.js:36]: setExam(exam);

File: src/app/exam/questions/ExamDescription.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    const [markdownText, setMarkdownText] = useState(undefined);
+    const [exam, setExam] = useState(undefined);
+    const [markdownTextBackUp, setMarkdownTextBackUp] = useState(undefined);
+
+    useEffect(() => {
+        if (!markdownText)
+            examService.getExam(examId).then(exam => {
+                setExam(exam);
+                setMarkdownText(exam.description);
+            });
```
Comment (Wally5077) [inline src/app/exam/questions/ExamDescription.js:24]: put ```then``` to next line

File: src/app/exam/options/ExamOptions.js [hunk 5]
```diff
@@ -9,117 +9,99 @@ import {examService} from "../../../services/services";
...
+                name: examName,
                 startTime,
                 endTime,
-                description: currentExam.description
-            }).then(() => {
-                refetchExam(examId)
-            })
-        }
+                description: exam.description
+            }).then(updateExams);
```
Comment (Wally5077) [inline src/app/exam/options/ExamOptions.js:47]: put ```then``` to next line

File: src/app/exam/options/ExamOptions.js [hunk 6]
```diff
@@ -9,117 +9,99 @@ import {examService} from "../../../services/services";
...
     };
 
     const deleteExam = () => {
-        examService.deleteExam(examId)
-            .then(() => {
-                    setShouldRedirect(true);
-                    setExams(exams.filter(g => g !== currentExam));
-                }
-            )
+        examService.deleteExam(examId).then(updateExams);
```
Comment (Wally5077) [inline src/app/exam/options/ExamOptions.js:52]: put ```then``` to next line

File: src/app/exam/participants/Examinees.js [hunk 7]
```diff
@@ -10,33 +9,28 @@ import {TextareaModal} from "../../commons/modals/TextareaModal";
...
     const [showAddStudentModal, setShowAddStudentModal] = useState(false);
     const [showAddGroupModal, setShowAddGroupModal] = useState(false);
     const [showRemoveExamineeConfirmationModal, setShowRemoveExamineeConfirmationModal] = useState(false);
 
-    const fetchExaminees = (examId) => {
-        examService.getExaminees(examId)
-            .then(examineeStudents => setExaminees(examineeStudents));
-    }
+    const fetchExaminees = useCallback(() => {
+        examService.getExaminees(examId).then(setExaminees);
```
Comment (Wally5077) [inline src/app/exam/participants/Examinees.js:25]: put ```then``` to next line

File: src/app/exam/participants/Examinees.js [hunk 8]
```diff
@@ -55,30 +49,23 @@ const Examinees = () => {
...
-    }
+                fetchExaminees();
+            });
+    };
 
     const addExamineesByGroups = async (groupNames) => {
-        examService.addGroupsOfExaminees(examId, groupNames)
-            .then(res => fetchExaminees(examId));
-    }
+        examService.addGroupsOfExaminees(examId, groupNames).then(fetchExaminees);
```
Comment (Wally5077) [inline src/app/exam/participants/Examinees.js:58]: put ```then``` to next line

File: src/app/exam/participants/Examinees.js [hunk 9]
```diff
@@ -55,30 +49,23 @@ const Examinees = () => {
...
+    };
 
     const removeExaminee = (email) => {
-        examService.deleteExaminees(examId, email)
-            .then(() => setExaminees(examinees.filter(examinee => examinee.email !== email)));
-    }
+        examService.deleteExaminees(examId, email).then(fetchExaminees);
```
Comment (Wally5077) [inline src/app/exam/participants/Examinees.js:62]: put ```then``` to next line

File: src/app/exam/score/ExamScorePage.js [hunk 10]
```diff
@@ -1,47 +1,38 @@
...
     const [presenter, setPresenter] = useState(undefined);
+    const [exam, setExam] = useState(undefined);
 
     useEffect(() => {
-        if (!currentExam) {
-            refetchExam(examId);
-        }
-    }, [currentExam, refetchExam, examId]);
+        if (!exam)
+            examService.getExam(examId).then(setExam);
```
Comment (Wally5077) [inline src/app/exam/score/ExamScorePage.js:19]: put ```then``` to next line

File: src/app/exam/scoreboard/ExamScoreboardPage.js [hunk 11]
```diff
@@ -1,51 +1,39 @@
...
+    const [exam, setExam] = useState(undefined);
     const [presenter, setPresenter] = useState(undefined);
 
     useEffect(() => {
-        if (!currentExam) {
-            refetchExam(examId);
-        }
-    }, [currentExam, refetchExam, examId]);
+        if (!exam)
+            examService.getExam(examId).then(setExam);
```
Comment (Wally5077) [inline src/app/exam/scoreboard/ExamScoreboardPage.js:21]: put ```then``` to next line

File: src/app/problem/ProblemList.js [hunk 12]
```diff
@@ -14,69 +15,63 @@ const ProblemList = () => {
     const history = useHistory();
     const [problems, setProblems] = useState(undefined);
     const [showCreateProblemModal, setShowCreateProblemModal] = useState(false);
-    const fetchProblems = useCallback(() => {
+    const fetchAllProblems = useCallback(() => {
         problemService.getAllProblems().then(setProblems);
```
Comment (Wally5077) [inline src/app/problem/ProblemList.js:19]: put ```then``` to next line

File: src/app/exam/submissions/ExamLiveSubmissions.js [hunk 13]
```diff
@@ -1,27 +1,25 @@
...
     let liveSubmissions = [];
 
     useEffect(() => {
-        if (!currentExam) {
-            refetchExam(examId);
-        }
-    });
+        if (!exam)
+            examService.getExam(examId)
+                       .then(setExam);
```
Comment (Johnny850807) [inline src/app/exam/submissions/ExamLiveSubmissions.js:21]: braces

File: src/app/exam/scoreboard/ExamScoreboardPage.js [hunk 14]
```diff
@@ -1,51 +1,39 @@
...
-                subscription.unsubscribe();
-            };
+            return () => subscription.unsubscribe();
         }
-    }, [currentExam, examId]);
-
+    }, [exam, examId, setPresenter]);
 
-    if (!currentExam || !presenter) {
+    if (!exam || !presenter)
```
Comment (Johnny850807) [inline src/app/exam/scoreboard/ExamScoreboardPage.js:32]: braces

File: src/app/exam/scoreboard/ExamScoreboardPage.js (hunk 11 above)
Comment (Johnny850807) [inline src/app/exam/scoreboard/ExamScoreboardPage.js:21]: braces

File: src/app/exam/score/ExamScorePage.js (hunk 10 above)
Comment (Johnny850807) [inline src/app/exam/score/ExamScorePage.js:19]: braces

File: src/app/exam/participants/Examinees.js [hunk 15]
```diff
@@ -10,33 +9,28 @@ import {TextareaModal} from "../../commons/modals/TextareaModal";
...
-            refetchExam(examId);
-        }
-        if (!examinees && currentExam) {
-            fetchExaminees(examId);
-        }
-    }, [currentExam, examId, examinees, refetchExam]);
+        if (!exam)
+            examService.getExam(examId)
+                .then(setExam)
+                .then(fetchExaminees);
```
Comment (Johnny850807) [inline src/app/exam/participants/Examinees.js:32]: braces

File: src/app/exam/options/ExamOptions.js [hunk 16]
```diff
@@ -9,117 +9,99 @@ import {examService} from "../../../services/services";
...
+        history.push("/exams");
     };
 
-    if (!currentExam) {
-        return <Spinner/>
-    }
+    if (!exam)
+        return <Spinner/>;
```
Comment (Johnny850807) [inline src/app/exam/options/ExamOptions.js:56]: braces

File: src/app/exam/options/ExamOptions.js (hunk 3 above)
Comment (Johnny850807) [inline src/app/exam/options/ExamOptions.js:36]: braces

File: src/app/exam/ExamList.js [hunk 17]
```diff
@@ -24,104 +25,103 @@ export const useExamList = () => {
...
-        setShouldRedirect(true);
+    const onExamCreated = exam => {
+        history.push(`/exams/${exam.id}/problems`);
     };
 
-    if (!exams || (shouldRedirect && !currentExam)) {
-        return <Spinner/>
-    }
+    if (!exams)
+        return <Spinner/>;
```
Comment (Johnny850807) [inline src/app/exam/ExamList.js:61]: braces

File: src/app/exam/ExamInPageNavigationBar.js [hunk 18]
```diff
@@ -1,28 +1,41 @@
...
-                setShouldRedirect(false);
+                history.push("/exams");
                 break;
             default:
                 console.log("WARNING: onBreadcrumbClickAtIndex index out of range.")
         }
     };
 
+    if (!exam)
+        return <></>;
```
Comment (Johnny850807) [inline src/app/exam/ExamInPageNavigationBar.js:32]: braces

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Braces

File: src/app/problem/editor/right/ProblemDescription.js [hunk 19]
```diff
@@ -42,7 +50,8 @@ const Description = () => {
                                          onClickEdit={onClickEdit}
                                          onClickSave={onClickSave}
                                          onClickCancel={onClickCancel}/>
-               }>
+               }
+               buttonPos="top">
```
Comment (Wally5077) [inline src/app/problem/editor/right/ProblemDescription.js:54]: Your Block property 'buttonPos' has default value 'top'

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

**Review by Johnny850807 — APPROVED** [review]
LGTM

File: src/app/exam/ExamInPageNavigationBar.js [hunk 20]
```diff
@@ -1,28 +1,42 @@
...
+const ExamInPageNavigationBar = () => {
+    const history = useHistory();
+    const {url} = useRouteMatch();
+    const {examId} = useParams();
+    const [exam, setExam] = useState(undefined);
+    const {shouldUpdate} = useExamContext();
+
+    useEffect(() => {
+        if (!exam || shouldUpdate) {
+            examService.getExam(examId).then(setExam);
```
Comment (Wally5077) [inline src/app/exam/ExamInPageNavigationBar.js:17]: put ```then``` to next line

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

**Review by Wally5077 — APPROVED** [review]
LGTM

**Pull request merged**